include/openssl/kdf.h

```diff
@@ -10,29 +10,135 @@
+
+# define EVP_KDF_CTRL_PASS                      (EVP_KDF_ALG_CTRL)
+# define EVP_KDF_CTRL_SALT                      (EVP_KDF_ALG_CTRL + 1)
+# define EVP_KDF_CTRL_ITER                      (EVP_KDF_ALG_CTRL + 2)
+# define EVP_KDF_CTRL_MD                        (EVP_KDF_ALG_CTRL + 3)
+# define EVP_KDF_CTRL_KEY                       (EVP_KDF_ALG_CTRL + 4)
+# define EVP_KDF_CTRL_TLS_SECRET                (EVP_KDF_ALG_CTRL + 5)
+# define EVP_KDF_CTRL_TLS_SEED                  (EVP_KDF_ALG_CTRL + 6)
+# define EVP_KDF_CTRL_HKDF_INFO                 (EVP_KDF_ALG_CTRL + 7)
+# define EVP_KDF_CTRL_HKDF_MODE                 (EVP_KDF_ALG_CTRL + 8)
+# define EVP_KDF_CTRL_SCRYPT_N                  (EVP_KDF_ALG_CTRL + 9)
+# define EVP_KDF_CTRL_SCRYPT_R                  (EVP_KDF_ALG_CTRL + 10)
+# define EVP_KDF_CTRL_SCRYPT_P                  (EVP_KDF_ALG_CTRL + 11)
+# define EVP_KDF_CTRL_SCRYPT_MAXMEM_BYTES       (EVP_KDF_ALG_CTRL + 12)
+
+# define EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND  0
+# define EVP_KDF_HKDF_MODE_EXTRACT_ONLY        1
+# define EVP_KDF_HKDF_MODE_EXPAND_ONLY         2
+
+# define EVP_KDF_CTX_set1_pass(kctx, pass, pwlen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_PASS, pwlen, (void *)(pass))
+
+# define EVP_KDF_CTX_set1_salt(kctx, salt, saltlen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_SALT, saltlen, (void *)(salt))
+
+# define EVP_KDF_CTX_set_iter(kctx, iter) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_ITER, iter, NULL)
+
+# define EVP_KDF_CTX_set_md(kctx, md) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_MD, 0, (void *)(md))
+
+# define EVP_KDF_CTX_set1_key(kctx, key, keylen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_KEY, keylen, (void *)(key))
+
+# define EVP_KDF_CTX_set1_tls1_prf_secret(kctx, sec, seclen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_TLS_SECRET, seclen, (void *)(sec))
+
+# define EVP_KDF_CTX_add1_tls1_prf_seed(kctx, seed, seedlen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_TLS_SEED, seedlen, (void *)(seed))
+
+# define EVP_KDF_CTX_add1_hkdf_info(kctx, info, infolen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_HKDF_INFO, infolen, (void *)(info))
+
+# define EVP_KDF_CTX_set_hkdf_mode(kctx, mode) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_HKDF_MODE, mode, NULL)
+
+# define EVP_KDF_CTX_set_scrypt_N(kctx, n) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_N, n)
+
+# define EVP_KDF_CTX_set_scrypt_r(kctx, r) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_R, r)
+
+# define EVP_KDF_CTX_set_scrypt_p(kctx, p) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_P, p)
+
+# define EVP_KDF_CTX_set_scrypt_maxmem_bytes(kctx, maxmem_bytes) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, \
+                             EVP_KDF_CTRL_SCRYPT_MAXMEM_BYTES, maxmem_bytes)
+
+
+/**** The old PKEY-based KDF API follows. ****/
```

Comment:
So these defines make the API source level compatible, but breaks the ABI, is this OK ?
Why not retain the original values so that ABI is not broken ?

Reply:
Some of the earlier constants have been merged in the new.  E.g. the salts.

A decision that the next OpenSSL release will be ABI breaking has been made, so this is okay.

Reply:
That shouldn't be too hard to fix.  Start with giving the new constants a different numeric range, then keep the old constants in an internal header file, renamed with `OLD_` in their names, and handle both the new and the old in the code where it makes sense.

Reply:
@paulidale, please note that formally speaking, we haven't actually made that decision yet.

Reply:
I am actually planning to backport this to 1.1, so it would be nice if there weren't gratuitous breakages if at all possible. I know how to handle this myself if you need to stick with the current code though, just mentioning why I care.

Reply:
If we rename the constants we're breaking the API -- the new names are different to the old.  The raised concern was that the constants were defined twice, rather than in terms of each other.

Back to defining them twice and staying ABI compatible, or define one in terms of the other ?

Reply:
I will put it back the way it was, preserving binary compatibility.

Reply:
I put the switch statement back that maps between the old and new cntrl codes and I moved the new codes to a different range so that it will force an error if the old constants are used with the new API.